plugins/catalog-backend/src/next/TaskPipeline.ts

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2021 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const DEFAULT_POLLING_INTERVAL_MS = 1000;
+
+type Options<T> = {
+  /**
+   * The callback used to load in new tasks. The number of items returned
+   * in the array must be at most `count` number of items, but may be lower.
+   *
+   * Any error thrown from this method fill be treated as an unhandled rejection.
+   */
+  loadTasks: (count: number) => Promise<Array<T>>;
+
+  /**
+   * The callback used to process a single item.
+   *
+   * Any error thrown from this method fill be treated as an unhandled rejection.
+   */
+  processTask: (item: T) => Promise<void>;
+
+  /**
+   * The target minimum number of items to process in parallel. Once the number
+   * of in-flight tasks reaches this count, more tasks will be loaded in.
+   */
+  lowWatermark: number;
+
+  /**
+   * The maximum number of items to process in parallel.
+   */
+  highWatermark: number;
+
+  /**
+   * The interval at which tasks are polled for in the background when
+   * there aren't enough tasks to load to satisfy the low watermark.
+   *
+   * @default 1000
+   */
+  pollingIntervalMs?: number;
+};
+
+/**
+ * Creates a task processing pipeline which continuously loads in tasks to
+ * keep the number of parallel in-flight tasks between a low and high watermark.
+ *
+ * @param options The options for the pipeline.
+ * @returns A stop function which when called halts all processing.
+ */
+export function startTaskPipeline<T>(options: Options<T>) {
+  const {
+    loadTasks,
+    processTask,
+    lowWatermark,
+    highWatermark,
+    pollingIntervalMs = DEFAULT_POLLING_INTERVAL_MS,
+  } = options;
+
+  if (lowWatermark >= highWatermark) {
+    throw new Error('lowWatermark must be lower than highWatermark');
+  }
+
+  let loading = false;
+  let stopped = false;
+  let inFlightCount = 0;
+
+  async function maybeLoadMore() {
+    if (stopped || loading || inFlightCount > lowWatermark) {
```

Comment:
```suggestion
    if (stopped || loading || inFlightCount >= lowWatermark) {
```

That's how I interpret the parameter

Reply:
Intention is that if you hit the low watermark, you load in more. It fits a bit better with for example 1/1, or even the 5/10 we're using, since otherwise you'd prolly want to set it to 6/10 as waiting until you hit 4 in-flight tasks is perhaps a bit low. Basically I like that that it's natural that the low watermark is half of the high one.

Reply:
Matter of perspective I guess, I really wanted to think of them in terms of the range of number of jobs you try to keep in flight, ie a min and a max. In the current design it's a bit like a min+1 and a max. Kinda suggested mine because it fit 1/1 better in my mind actually :) Because the 1 state is then the_stable_ one. Then when that one is finished you hit 0 which is below the mark and immediately try to load a new one 

Anyway this is super nitpicky